### ISSUE TITLE
constant/generic/global/local objects

### ISSUE THREAD

**Issue by airlied** [triSYCL/triSYCL#174]
In address_space.hpp we define

cl::sycl::constant<T>
cl::sycl::generic<T>
etc.

alongside the cl::sycl::constant_ptr<T> ones, however in the spec I can't see any mention of these, (the non-ptr variants), have they been removed from the spec? A few tests use them so we'd have to fix that first if so.

**Comment by keryell** [comment]
It is some experimenting ahead with explicit constructs that could be used in SYCL Next for in-kernel or program-scope variables with explicit address space.
So not it the current standard, but triSYCL is an open-source playground to experiment publicly, so I would just keep it as is. But we should be probably clearer in the tests or in the implementation about specifying that some things are experimental and not in the standard.

**Comment by airlied** [comment]
Cool I think we probably need to add a TRISYCL_FUTURE or TRISYCL_EXPERIMENTAL define so we can carve these out while we progress towards the 1.2.1 conformance.

**Comment by keryell** [comment]
But since there is no negative testing in Khronos conformance tests, how can it impact the CTS result?

There should be no test for example to know whether there are some other features outside of the basic implementation or to check explicitly than something outside of the spec is really failing to compile or to run. Otherwise it would be difficult to have extensions or to hide implementation details.

Have you found a test which breaks because there are more things in triSYCL than in the specification?

I expect it is only the opposite: most of the tests breaking because triSYCL implements right now quite less than what is defined in the specification... :-/

**Comment by airlied** [comment]
Well the problem is with things where the spec has diverged from the SYCL baseline but then we've enhanced the baseline, like for example these objects. The spec for global_ptr and all related _ptr has changed from what we have in trisycl, fixing that means changing a bunch of stuff around this area, but I then have to figure out how to fix up a bunch of code and tests but have no spec on what they are trying to aim for.

**Comment by airlied** [comment]
like boost::experimental namespace btw

**Comment by keryell** [comment]
Yes you are right, it is difficult to have the big picture with many years of loose co-development of the specification and a partial implementation...
